plugins/newsql/newsql.c

```diff
@@ -1213,23 +1213,21 @@ static int newsql_param_value(struct sqlclntstate *clnt,
     param->name = val->varname;
     param->pos = val->has_index ? val->index : 0;
     param->type = newsql_to_client_type(val->type);
-
+    int len = val->value.len;
     void *p = val->value.data;
 
-    if (val->has_isnull && val->isnull) {
+    /* The bound parameter is from an old client which does not send isnull,
+       and its length is 0. Treat it as a NULL to keep backward-compatible. */
+    if (len == 0 && !val->has_isnull) {
         param->null = 1;
         return 0;
     }
 
-    if (val->value.data == NULL) {
-        if (param->type != CLIENT_BLOB) {
-            param->null = 1;
-            return 0;
-        }
-        p = (void *)"";
+    if (val->isnull) {
```

Comment:
Also test for `val->has_isnull` ?

Reply:
`val->isnull` being true implies `val->has_is_null` being true too.

Reply:
Basically `isnull` is tri-state.

| state | has_isnull | isnull |
| - | - | - |
| not applicable | 0 | 0 |
| NULL | 1 | 1 |
| 0-length | 1 | 0 |

Reply:
I was wondering if it's safe to read 'isnull' when 'has_isnull' is false. Turns out protobuf initializes the field with the default value. So, should be good. Thanks for clarifying.

Reply:
Anytime!